tests/e2e/tests/mixer/mixer_test.go

```diff
@@ -311,53 +281,47 @@ func TestRateLimit(t *testing.T) {
 		}
 	}()
 
-	// allow time for configuration to go and be active.
-	// TODO: figure out a better way to confirm rule active
-	time.Sleep(1 * time.Minute)
-	threads := 5
-	timeout := "10s"
-	connections := 15
-	duration := "2m"
-	errorFile := filepath.Join(tc.Kube.TmpDir, "TestRateLimit.err")
-	JSONFile := filepath.Join(tc.Kube.TmpDir, "TestRateLimit.Json")
-	url := fmt.Sprintf("%s/productpage", tc.gateway)
+	// allow time for rule sync
+	time.Sleep(10 * time.Second)
 
-	luaScript := &util.LuaTemplate{
-		TemplatePath: util.GetResourcePath(luaTemplate),
-		Template: &luaScriptTemplate{
-			ErrorFile: errorFile,
-			JSONFile:  JSONFile,
-		},
-	}
-	if err := luaScript.Generate(); err != nil {
-		t.Errorf("Failed to generate lua script. %v", err)
-		return
+	url := fmt.Sprintf("%s/productpage", tc.gateway)
+	opts := fortio.DefaultRunnerOptions
+	opts.Duration = 30 * time.Second
+	opts.QPS = 30
+	clients := make([]fortio.Fetcher, 0, opts.NumThreads)
+	for i := 0; i < opts.NumThreads; i++ {
+		clients = append(clients, fortio.NewBasicClient(url, "1.1", true))
 	}
 
-	if err := tc.wrk.Run(
-		"-t %d --timeout %s -c %d -d %s -s %s %s",
-		threads, timeout, connections, duration, luaScript.Script, url); err != nil {
-		t.Errorf("Failed to run wrk %v", err)
-		return
-	}
+	var totalReqs, successReqs, errReqs int64
 
-	// TODO: Use the JSON file to do more detailed checking
-	r := &luaScriptResponseJSON{}
-	if err := util.ReadJSON(JSONFile, r); err != nil {
-		t.Errorf("Could not parse json. %v", err)
-		return
-	}
+	opts.Function = func(tid int) {
+		code, _, _ := clients[tid].Fetch()
+		atomic.AddInt64(&totalReqs, 1)
```

Comment:
any chance to use a context or per thread/goroutine for that instead of atomic ?

like https://github.com/istio/istio/blob/master/devel/fortio/cmd/fortio/main.go#L48

(maybe I should add stuff to make it easier to reuse and aggregate without copy pasting)

Reply:
the layer etc is done as part of #519